meson: Refine GLib requirement definition

tintou opened the pull request:
Set all GLib libraries as requiring the same version, always keep the MIN/MAX macro definition as the same as the version requirement.

ximion commented:
The CI is rather unhappy, will you modernize the sources for the GLib requirements as well? We can also drop any superseded version checks then, if any still remain.

sidt4 commented:
`GLIB_VERSION_MIN_REQUIRED` is set to 2.44 to avoid deprecation warnings / errors which are yet to be fixed in code.

I think we bump the `GLIB_VERSION_MIN_REQUIRED` value in stages once the depredations are fixed in code.

tintou commented on meson.build:
```diff
@@ -64,8 +69,11 @@ endif
-add_project_arguments ('-DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_44', language: 'c')
+add_project_arguments (
+  '-DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_@0@_@1@'.format(glib_major, glib_minor),
+  '-DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_@0@_@1@'.format(glib_major, glib_minor),
```
Setting the MAX_ALLOWED macro allows to have a warning when using something that is only accessible in newer GLib, leaving it out for now

sidt4 replied:
I think it makes sense to have `GLIB_VERSION_MAX_ALLOWED` here. This will signal incorrect glib deps version in `meson.build`. For instance, if I've not bumped glib version to `2.70` in https://github.com/PackageKit/PackageKit/pull/842, the code will fail to build (if the following patch was in `meson.build`). 

```diff
--- a/meson.build
+++ b/meson.build
@@ -66,6 +66,7 @@ add_project_arguments ('-D_DEFAULT_SOURCE', language: 'c')
 
 # Avoid g_simple_async_result deprecation warnings in glib 2.46+
 add_project_arguments ('-DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_44', language: 'c')
+add_project_arguments ('-DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_62', language: 'c')
 
 # allow the daemon to include library files directly
 add_project_arguments ('-DPK_COMPILATION', language: ['c', 'cpp'])
```

### Compilation error:

```
./lib/packagekit-glib2/pk-debug.c: In function ‘pk_debug_pre_parse_hook’:
../lib/packagekit-glib2/pk-debug.c:143:27: error: Not available before 2.70 [-Werror]
  143 |                 G_OPTION_ENTRY_NULL
      |                           ^~~~~~~~~                  
cc1: all warnings being treated as errors
```

tintou replied:
The issue I had is that we have a:
```
#if GLIB_CHECK_VERSION(2, 70, 0)
		g_source_set_static_name (source, "PkTransaction progress timeout");
#endif
```
So I prefer to leave it out of this MR for now, and add it once we require GLib 2.70

sidt4 replied:
https://github.com/PackageKit/PackageKit/pull/842 requires GLib 2.70, so I think we can soon bump to 2.70.

ximion replied:
I'm definitely not against raising the GLib baseline version, but let's do that in another PR :-)

tintou commented:
Alright, added several commits to modernize the codebase

ximion commented:
Oooof, thank you so much for all the work you're doing! I also apologize in advance, this will take me a bit of time to review for sure. I also want to make sure we're not breaking ABI by accident with this (if it's unavoidable, we can think about that though).

sidt4 commented:
Many thanks for looking into this :1st_place_medal:.

I was planning to take a look into this while fixing https://github.com/PackageKit/PackageKit/pull/842, but happy it's done. 

Any chance we can have smaller PRs (if that'll help @ximion to do a quicker review).

- PR1 (code changes for `DGLIB_VERSION_MIN_REQUIRED=2.50`)
- PR2 (code changes for `DGLIB_VERSION_MIN_REQUIRED=2.56`)
- PR3 (code changes for `DGLIB_VERSION_MIN_REQUIRED=2.58`)
- PR4 (code changes for `DGLIB_VERSION_MIN_REQUIRED=2.62`)

etc.

That way we can have the PRs committed quickly in a span of a week than a single huge merge. Again, please wait for confirmation from `ximion` before spending time splitting MRs, as it's ultimately his call.

tintou commented:
I intentionally did minimal changes to the lib/ folder so that the ABI remains untouched

sidt4 commented:
>  also want to make sure we're not breaking ABI by accident with this (if it's unavoidable, we can think about that though).

I remember some GNOME / FreeDesktop library doing automatic ABI check as part of build.

Is it possible to do that in `PackageKit`?

sidt4 commented:
> I remember some GNOME / FreeDesktop library doing automatic ABI check as part of build.

Something like https://github.com/GNOME/grilo/commit/e9d6f666ce2da99bd3994d0fe54c8593c0906058.

tintou commented:
@sidt4 done in https://github.com/PackageKit/PackageKit/pull/859 ;)

ximion reviewed (changes requested):
I finally had time for a review, and this change looks good! This was mind-numbing to look at, but I'm not complaining, thank you for doing what is basically maintenance work!

I have one request: Could you please define a `GET_PRIVATE` macro for the `*_get_instance_private` queries and use that to fetch `priv`? That would make PK consistent with other code using this pattern and other parts of the codebase that also use this. Example: `#define GET_PRIVATE(o) (pk_results_get_instance_private (o))` (right after `G_DEFINE_TYPE_WITH_PRIVATE`).

It's nitpick-y, but I find this style very useful to simplify GObject code and to prevent future errors :-)

Also, some apology in advance, but I will assign copilot to "review" this as well - I have never done this because I severely doubt it's usefulness, but for a change like this that is fairly straight-forward but touches a lot of files, it may actually be beneficial and maybe might catch some stuff that my manual review has missed.

In case it spams nonsense, I'll deal with that, no worries! :-)
Thank you very much for this work!

pwithnall commented:
> I think confirmation from @pwithnall for comments from @tintou settles the GNOME part.
> 
> Some comments from Philip would be helpful to resolve this blocking topic.

I’m not sure exactly what you’re asking so I’ll just give some clarification over my thumbs-up emoji reactions above.

I thumbs-up-ed Corentin’s comments because he’s right that `GET_PRIVATE()` is a remnant from before `*_get_instance_private()` inline methods were generated. My preferred coding style does not use `GET_PRIVATE()`, it call the `*_get_instance_private()` method directly. But it’s also just a coding style preference and imo not at all worth blocking this PR over, one way or the other.

tintou commented:
I indeed prefer `*_get_instance_private()` but I'm not opposed to anything else if one decide that it should be the codestyle in use, it is just that it requires to revert a lot of changes here 😅

ximion commented:
I do prefer `GET_PRIVATE()` with my maintainer hat on for PK, *but* it really is just a light style preference because it's already used in PK as well as in other GObject projects I work on (and once we can depend on C23 we can make it so that the type of the object doesn't appear at all anymore when fetching `priv`, which would be quite neat, less typing and verbosity).

@tintou no need to revert and redo all the commits, you could make them all at once in one commit, that would be fine with me.
I could also make that change happen later in a follow-up commit, if you don't want to do it, that's also perfectly okay :-)

sidt4 commented:
@tintou: can you take a look at the above issue?

I've installed `PackageKit` main + this MR + fix on my Debian Unstable system. I'm using it since yesterday without issues. Tried install / remove / offline update etc.

Also, it would be nice to get this MR committed (than blocking on the `GET_PRIVATE()` issue), so that we can capture any issues (like the above comment) sooner.

tintou commented:
> I encounter crashes, as `G_DEFINE_TYPE` is not replaced with `G_DEFINE_TYPE_WITH_PRIVATE` for the following types.
> 
> ```
> PkDesktop
> PkDetails
> ```

Thanks! I've added G_DEFINE_TYPE_WITH_PRIVATE to PkDetails, PkDesktop has no use of the PkDesktopPrivate struct anymore (also removed now while keeping it defined in the header for API compatibility).

tintou commented:
> I am only waiting for a response / indication of preference by @tintou on:
> 
> > @tintou no need to revert and redo all the commits, you could make them all at once in one commit, that would be fine with me.
> > I could also make that change happen later in a follow-up commit, if you don't want to do it, that's also perfectly okay :-)
> 
> ...well, and at this point for a rebase/conflict resolution. And then I would also prefer to merge this ASAP, because keeping large MRs around that touch a lot of files at once is rather painful for everyone involved (including other pending MRs that need rebasing once this is merged).

I don't have the bandwidth to do the switch back to GET_PRIVATE commit soon so If you've some time to do it that would be great!
